### PR TITLE
mlxrunner: add structured output support

### LINKED ISSUE #16563: Structured outputs appear to be ignored for MLX models
### What is the issue?

Environment:

Ollama 0.30.6
macOS
Reproduced with both Qwen 3.5 and Gemma 4

Minimal repro:

import httpx

SCHEMA = {
    "type": "object",
    "properties": {
        "animal": {"type": "string"},
    },
    "required": ["animal"],
    "additionalProperties": False,
}

for model in [
    "qwen3.5:4b",
    "qwen3.5:4b-mlx",
]:
    r = httpx.post(
        "http://localhost:11434/api/chat",
        json={
            "model": model,
            "messages": [{"role": "user", "content": "Say hi"}],
            "stream": False,
            "format": SCHEMA,
        },
        timeout=300,
    )
    print(model)
    print(r.json()["message"]["content"])

Observed output:

qwen3.5:4b
{"animal": "👋 Hey there! Hello! How can I help you today? 😊"}

qwen3.5:4b-mlx
Hi there! 👋 How's it going? Anything I can help you with today?

Expected behavior:

Both models should produce output conforming to the supplied schema.

Actual behavior:

The non-MLX model appears to be influenced by or constrained by the schema. The MLX model appears to ignore the schema entirely and responds naturally to the prompt.

I see the same pattern with another model family:

gemma4:31b
-> schema-conforming JSON

gemma4:31b-mlx
-> natural-language greeting

This suggests the issue may be specific to the MLX execution path rather than a particular model family.

### Relevant log output

```shell
$ python scripts/minimal-repro.py
Ollama version: 0.30.6

Prompt: "Say hi"

Schema:
{'type': 'object', 'properties': {'animal': {'type': 'string'}}, 'required': ['animal'], 'additionalProperties': False}

Model: qwen3.5:4b
Output: {"animal": "👋 Hey there! Hello! How can I help you today? 😊"}

Model: qwen3.5:4b-mlx
Output: Hi there! 👋 How's it going? Anything I can help you with today?

Model: gemma4:31b
Output: {
  "animal": "dog owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns owns Owns"
}


Model: gemma4:31b-mlx
Output: Hi!

--

Exact script used for local repro:

import httpx

SCHEMA = {
    "type": "object",
    "properties": {
        "animal": {"type": "string"},
    },
    "required": ["animal"],
    "additionalProperties": False,
}

PROMPT = "Say hi"

MODELS = [
    "qwen3.5:4b",
    "qwen3.5:4b-mlx",
    # Uncomment if you want a second model family:
    "gemma4:31b",
    "gemma4:31b-mlx",
]


def run_model(model: str) -> None:
    response = httpx.post(
        "http://localhost:11434/api/chat",
        json={
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": PROMPT,
                }
            ],
            "stream": False,
            "format": SCHEMA,
        },
        timeout=900,
    )

    response.raise_for_status()
    payload = response.json()

    print(f"\nModel: {model}")
    print(f"Output: {payload['message']['content']}")


def main() -> None:
    version = httpx.get("http://localhost:11434/api/version").json()
    print(f"Ollama version: {version['version']}")
    print(f'\nPrompt: "{PROMPT}"')
    print(f"\nSchema:\n{SCHEMA}")

    for model in MODELS:
        run_model(model)


if __name__ == "__main__":
    main()
```

### OS

macOS

### GPU

Apple

### CPU

Apple

### Ollama version

0.30.6

### PR DESCRIPTION
**Pull request by jessegross**
The MLX runner accepted the API's `format` field but did not enforce it: requests asking for JSON or a JSON Schema got unconstrained text, and clients had no way to tell.

This enforces `format` with xgrammar. Each sampling step masks the logits to the tokens the grammar allows, so every emitted token and the end of generation are valid under the constraint. Sampling, penalties, and logprobs see the constrained distribution, and `"json"` yields a JSON object, as the API documents and the llama-server path already enforces. Only sampling waits on the mask; the forward pass is dispatched before it, so constrained decoding stays pipelined.

The grammar engine is a dynamic library alongside MLX: when it is missing, plain inference is unaffected and structured requests fail with an explicit error rather than returning unconstrained text.

What this does not cover:
- Constrained requests decode without speculative decoding for now.
- xgrammar enforces a JSON Schema subset. Most gaps drop a single keyword's constraint; multi-branch `allOf` degrades to accepting any JSON value, matching what vllm and sglang ship with the same engine.

Decoding 256 tokens of a book-list schema on qwen3.8:27b-mlx (M5 Max, seed 42, thinking off); pre-decode is the request time spent before the first token:

| configuration | decode | pre-decode |
|---|---|---|
| unconstrained | ~65 tok/s | ~70 ms |
| unconstrained, no draft | ~32 tok/s | ~70 ms |
| JSON schema | ~32 tok/s | ~70 ms |

Schema and draft-less decoding are equal to within 0.1 tok/s in paired adjacent requests, and a cold grammar compile adds nothing measurable to pre-decode. The gap to unconstrained decoding is the disabled draft model.

Two preparatory commits are included: `llm` drops the unused `Grammar` completion-request field, and the `mlx` value readers now return exact element types and evaluate their arrays before reading — the latter closes an unsynchronized read the grammar path would otherwise hit when reading committed tokens for the matcher.

Fixes #16563

### CODE REVIEW

**Comment by rick-github** [comment]
Am I correct in thinking that thinking output (if any) will also be constrained to the grammar?

**Review by dhiltgen — APPROVED** [review]

**Comment by jessegross** [comment]
> Am I correct in thinking that thinking output (if any) will also be constrained to the grammar?

The Ollama server currently breaks up the request internally to thinking and non-thinking pieces to avoid that problem, with only the non-thinking portions being constrained. The runner code here provides the basis to handle that split more natively (though it just matches the current runner API for now) and also go further to enforce specific schemas for different tools.

**Comment by trebor** [comment]
thank you!  this has been a significant hang up for me!
